incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -59,51 +59,121 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 		if errors.IsNotFound(err) {
-			// If the instance doesn't exist, return nil to prevent a retry.
 			return ctrl.Result{}, nil
 		}
 		return ctrl.Result{}, err
 	}
 
-	// Report "Forbidden" state and early exist if the namespace is not allowed to self-serve
-	// namespaces but has bypassed the webhook and successfully created the hns instance.
+	// Report "Forbidden" state and early exist if the namespace is not allowed to
+	// self-serve namespaces but has bypassed the webhook and successfully created
+	// the hns instance. Forbidden HNSes won't have finalizers.
 	// TODO refactor/split the EX map for 1) reconciler exclusion and 2) self-serve not allowed
 	//  purposes. See issue: https://github.com/kubernetes-sigs/multi-tenancy/issues/495
 	if EX[pnm] {
 		inst.Status.State = api.Forbidden
 		return ctrl.Result{}, r.writeInstance(ctx, log, inst)
 	}
 
-	// Get the self-serve namespace's hierarchyConfig and namespace instances.
-	nsInst, err := r.getNamespace(ctx, nm)
+	// Get the owned child namespace instance. If it doesn't exist, initialize one.
+	cnsInst, err := r.getNamespace(ctx, nm)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	switch {
-	case nsInst.Name == "":
-		log.Info("The owned subnamespace does not exist", "subnamespace", nm)
-		inst.Status.State = api.Missing
+	if deleting, err := r.onDeleting(ctx, log, inst, cnsInst); deleting {
+		return ctrl.Result{}, err
+	}
+
+	// Update the state. If the owned child namespace doesn't exist, create it.
+	// (This is how a self-service namespace is created through a CR)
+	r.updateState(log, inst, cnsInst)
+	if inst.Status.State == api.Missing {
 		if err := r.writeNamespace(ctx, log, nm, pnm); err != nil {
 			return ctrl.Result{}, err
 		}
-	case nsInst.Annotations[api.AnnotationOwner] != pnm:
-		log.Info("The owner annotation of the subnamespace doesn't match the owner", "annotation", nsInst.Annotations[api.AnnotationOwner])
+	}
+
+	// Add finalizers on all non-forbidden HNSes to ensure it's not deleted until
+	// after the owned child namespace is deleted.
+	inst.ObjectMeta.Finalizers = []string{api.MetaGroup}
+	return ctrl.Result{}, r.writeInstance(ctx, log, inst)
+}
+
+func (r *HierarchicalNamespaceReconciler) onDeleting(ctx context.Context, log logr.Logger, inst *api.HierarchicalNamespace, cnsInst *corev1.Namespace) (deleting bool, err error) {
+	// Early exit and continue reconciliation if the instance is not being deleted.
+	if inst.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	log.Info("The HNS instance is being deleted.")
+	switch {
+	case r.doNothing(log, inst, cnsInst):
+		log.Info("Do nothing in this reconciliation.")
+		return true, nil
+	case r.removeFinalizers(log, inst, cnsInst):
+		log.Info("Remove the finalizers.")
+		return true, r.writeInstance(ctx, log, inst)
+	default:
+		// The owned namespace is not being deleted and it allows cascadingDelete.
+		// Delete the owned namespace.
+		return true, r.deleteNamespace(ctx, log, cnsInst)
+	}
```

Comment:
Nit: I feel like this might be easier to read in the opposite direction:

* Is there a child namespace that isn't being deleted? Delete it.
* If there's no namespace, are there finalizers? If so, remove them.
* Default: there must not be anything to do - wait for something to happen (the default case is usually the do-nothing case).

This (roughly) matches the order in which things will actually happen too.

Reply:
sg, will update it.